Add FNV-1a hash functions

dwmw2 opened the pull request:
Add `fnv1a_hash()` and also `fnv1a_hash_extend()` which takes an existing hash value and continues the FNV-1a hash calculation with additional string data. This allows building up a hash incrementally.

Both functions have `const char*` and `const std::string& overloads`, like the legacy `fnv1_hash()` (which is untouched by this PR).

# What does this implement/fix?

This adds the ability to extend a hash with an additional string. We'll use this for hashing the build info *plus* a component-specific thing like a serial number.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840
- [x] host

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

dwmw2 committed:
Add fnv1_hash_extend() function to continue FNV-1 hashing

Add fnv1_hash_extend() which takes an existing hash value and continues
the FNV-1 hash calculation with additional string data. This allows
building up a hash incrementally.

Refactor fnv1_hash() to be an inline wrapper around fnv1_hash_extend()
with the FNV-1 initial offset basis, eliminating code duplication.

Both functions have const char* and const std::string& overloads.

dwmw2 committed:
Add component test for fnv1_hash and fnv1_hash_extend

Add test that verifies:
- fnv1_hash() produces correct hash values
- fnv1_hash_extend() correctly continues hashing
- Extending a hash produces same result as hashing concatenated string
- Both const char* and std::string overloads work correctly

Test runs on boot via lambda and logs results.

dwmw2 committed:
Merge branch 'dev' into extend-hash

bdraco commented:
I haven't looked at the contents but I think one thing to point out is the recall the hash in python FNV1a and the hash in cpp might be standard fnv1

dwmw2 commented:
> I haven't looked at the contents but I think one thing to point out is the recall the hash in python FNV1a and the hash in cpp might be standard fnv1

Oh, joy. You're right. In that case I'll redo this and add fnv1a_hash() and fnv1a_extend() to the C++ side. Unless you'd prefer that I switch to using FNV1 in the buildinfo config_hash?

dwmw2 committed:
Merge branch 'extend-hash' of github.com:dwmw2/esphome into extend-hash

dwmw2 committed:
Revert "Add component test for fnv1_hash and fnv1_hash_extend"

This reverts commit 62c70b391b4cb945326070ab1ed2b8012197ca62.

dwmw2 committed:
Revert "Add fnv1_hash_extend() function to continue FNV-1 hashing"

This reverts commit c014046edf6b79c02f559c7c160d7ffe15637ddc.

dwmw2 committed:
Add FNV-1a hash functions

Add fnv1a_hash() and fnv1a_hash_extend() as preferred alternatives to
the legacy fnv1_hash(). FNV-1a uses XOR-then-multiply instead of
multiply-then-XOR, providing better avalanche characteristics.

The implementation supports both C strings and std::string, with
fnv1a_hash_extend() allowing incremental hashing of multiple strings.

dwmw2 committed:
Add integration test for FNV-1a hash functions

Test validates:
- Known FNV-1a hash values against specification
- fnv1a_hash_extend() produces same result as single hash
- std::string overloads work correctly

dwmw2 committed:
Merge branch 'dev' of github.com:esphome/esphome into extend-hash

bdraco commented on tests/integration/test_fnv1a_hash.py:
```diff
@@ -0,0 +1,59 @@
+        assert device_info.name == "fnv1a-hash-test"
+
+        # Wait for boot to complete
+        await asyncio.sleep(2.0)
```
Best to use a future to wait till its logged so we don't have lots of wait time in tests (integration tests already take a long time and its starting to be a problem)

Example in the README.md

dwmw2 replied:
Bah, meant to fix that before pushing. Thanks for spotting it.

dwmw2 committed:
Complete FNV-1a test immediately when all results seen

Use asyncio.Event to exit as soon as all 4 expected test results are
captured, rather than waiting 2 seconds unconditionally. Timeout after
2 seconds if not all tests complete, showing which tests were found.

bdraco commented on esphome/core/helpers.h:
```diff
@@ -378,9 +378,19 @@ uint16_t crc16be(const uint8_t *data, uint16_t len, uint16_t crc = 0, uint16_t p
+  return fnv1a_hash_extend(hash, str.c_str());
+}
+/// Calculate a FNV-1a hash of \p str.
+inline uint32_t fnv1a_hash(const char *str) { return fnv1a_hash_extend(2166136261UL, str); }
```
please make `2166136261UL` a constexpr .. maybe `FNV1A_OFFSET_BASIS`

bdraco commented on tests/integration/fixtures/fnv1a_hash.yaml:
```diff
@@ -0,0 +1,52 @@
+        using esphome::fnv1a_hash_extend;
+
+        // Test known FNV-1a hashes
+        uint32_t hash_hello = fnv1a_hash("hello");
```
we have had issue with fnv1_hash("") before so we should get a test for fnv1a_hash("");

dwmw2 committed:
Use named constants for FNV-1 offset basis and prime

Replace magic numbers with FNV1_OFFSET_BASIS and FNV1_PRIME constexpr
values for better code clarity.

dwmw2 committed:
Add test for FNV-1a hash of empty string

Verify that hashing an empty string returns the offset basis value.

bdraco reviewed (approved):
Thanks @dwmw2
